### PR TITLE
Add script for basic memory pinning CUDA benchmark

### PR DESCRIPTION
**Pull request by avivhaber**
Signed-off-by: Aviv Haber <aviv@anyscale.com>

## Why are these changes needed?

Script for benchmarking memory pinning on pytorch. The program trains one of two models on CIFAR10 and reports the time taken loading and training.

The first model (small) is a simple CNN described in [this tutorial](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html).  The second model (large) is the [torch VGG11 model](https://pytorch.org/vision/main/models/generated/torchvision.models.vgg11.html).

## Usage
The timer starts after the dataset has been loaded and the model has begun training.

To run the benchmark:
`python torch_test.py --model=simplecnn --pin`
or `python torch_test.py --model=simplecnn`

You can also use `--model=vgg11`. This is a larger model and will take longer. 

The `--pin` option determines whether to use memory pinning.

You can select the dataset with `--dataset=imagenet` or `--dataset=cifar10`. The default is CIFAR. If you use imagenet you must also pass `--imagenetpath=/path/to/imagenet/root`.

You can use `--size=512` to transform the images to 512x512, for instance. The default for CIFAR10 is 32x32. The default for ImageNet is 256x256.

## Results \# 1
Following are typical results (little change between runs). Default 32x32 images, CIFAR10.
Code was run on a single g4dn.4xlarge node. `BATCH_SIZE=4, NUM_WORKERS=0, EPOCHS=1`
| | Small Model | Large Model|
|---|---|---|
| Pinning off | 41s | 469s |
| Pinning on | 41s | 466s |

There is little to no difference. Maybe this is because the node has enough system memory that it rarely needs to spill to disk when pinning is off. I'm going to run more benchmarks on larger datasets

## Results \# 2
Using the same CIFAR10 dataset, but transforming the images to be 512x512 (Use `--size=512`), we get the following results. `BATCH_SIZE=100, NUM_WORKERS=8, EPOCHS=1`
| | Small Model | Large Model|
|---|---|---|
| Pinning off | 354s |  |
| Pinning on | 334s |  |

Here with the larger tensor size, the pinning actually has an effect (~6% difference).

## Results \# 3
Using a small subset of imagenet and only doing training (no testing). The images were transformed to 256x256. `BATCH_SIZE=100, NUM_WORKERS=8, EPOCHS=5`
| | Small Model | Large Model|
|---|---|---|
| Pinning off | 40s | 217s  |
| Pinning on | 35s | 203s |

~6% difference for large model, ~13% difference for small model.

### CODE REVIEW

**Comment by clarng** [comment]
Btw can we target the merge to a non-master branch, and use that branch for all the experiments of this project

File: python/ray/experimental/gpu_buffer/torch_test.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    total = 0
+    correct = 0
+    with torch.no_grad():
+        for data in testloader:
+            images, labels = data[0].to(device, non_blocking=pinned), data[1].to(device, non_blocking=pinned)
+            outputs = net(images)
+            _, predicted = torch.max(outputs.data, 1)
+            total += labels.size(0)
+            correct += (predicted == labels).sum().item()
+    return correct / total
```
Comment (cadedaniel) [inline python/ray/experimental/gpu_buffer/torch_test.py:43]: I would need to check this; line 42 might be done on CPU, which causes the GPU to block until the CPU computation is done. This would negate any benefits of nonblocking transfers.

Comment (cadedaniel) [reply python/ray/experimental/gpu_buffer/torch_test.py:43]: In general we want the CPU to run as far ahead as possible of the GPU, so that the GPU remains saturated.

Comment (avivhaber) [reply python/ray/experimental/gpu_buffer/torch_test.py:43]: Using `tensor.get_device()`, I verified that `outputs`, `predicted`, and `labels` are on the GPU. Is that enough to conclude line 42 is done on GPU?

Comment (cadedaniel) [reply python/ray/experimental/gpu_buffer/torch_test.py:43]: [`.item()`](https://pytorch.org/docs/stable/generated/torch.Tensor.item.html) will convert the tensor value to a Python value, which blocks the CPU until the computation is complete

**Comment by jiaodong** [comment]
Good findings, later you can also compare this with AIR's data loading such as our benchmark: https://sourcegraph.com/github.com/ray-project/ray/-/blob/release/air_tests/air_benchmarks/workloads/pytorch_training_e2e.py?L51

Note that AIR doesn't use pytorch dataloader and use `iter_torch_batches` instead, which will behave quite differently compare to pytorch.

File: python/ray/experimental/gpu_buffer/torch_test.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    parser.add_argument(
+        '--model',
+        help='Which model to train. Small is simple CNN, large is VGG11.',
+        choices=['small', 'large'],
+        default='large',
+        required=False,
+    )
+    parser.add_argument(
+        '--size',
+        help='Width/height to scale the images to. Only applies to CIFAR10 dataset. Prefer a power of 2.',
```
Comment (clarng) [inline python/ray/experimental/gpu_buffer/torch_test.py:46]: only working for one dataset seems a little non-intuitive. Allow each dataset to have its default size and make this param optional?

**Review by clarng — COMMENTED** [review]
looks good so far! A few nits

**Review by clarng — APPROVED** [review]
looks good!
